include/nuttx/net/netdev_lowerhalf.h

```diff
@@ -0,0 +1,402 @@
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_NET_NETDEV_LOWERHALF_H
+#define __INCLUDE_NUTTX_NET_NETDEV_LOWERHALF_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <net/if.h>
+#include <netinet/in.h>
+#include <stdatomic.h>
+
+#include <nuttx/net/ip.h>
+#include <nuttx/net/net.h>
+#include <nuttx/net/netdev.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* | <-------------- NETPKT_BUFLEN ---------------> |
+ * +---------------------+-------------------+------+      +-------------+
+ * | reserved for driver |       data        | free | ---> | next netpkt |
+ * +---------------------+-------------------+------+      +-------------+
+ * |                     | <--- datalen ---> |
+ * ^base                 ^data
+ */
+
+#define NETPKT_BUFLEN   CONFIG_IOB_BUFSIZE
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* We use IOB as our buffer now, we may change to some other structure when
+ * needed, so define a type netpkt_t for lower half.
+ * TODO: Provide interface of its queue, maybe a simple wrapper of iob_queue.
+ */
+
+typedef struct iob_s netpkt_t;
+
+enum netpkt_type_e
+{
+  NETPKT_TX,
+  NETPKT_RX,
+  NETPKT_TYPENUM
+};
+
+/* This structure is the generic form of state structure used by lower half
+ * netdev driver. This state structure is passed to the netdev driver when
+ * the driver is initialized. Then, on subsequent callbacks into the lower
+ * half netdev logic, this structure is provided so that the netdev logic can
+ * maintain state information.
+ *
+ * Normally that netdev logic will have its own, custom state structure
+ * that is simply cast to struct netdev_lowerhalf_s. In order to perform such
+ * casts, the initial fields of the custom state structure match the initial
+ * fields of the following generic netdev state structure.
+ */
+
+struct netdev_ops_s;
+struct netdev_lowerhalf_s
+{
+  FAR const struct netdev_ops_s *ops;
+  atomic_int quota[NETPKT_TYPENUM]; /* Max # of buffer held by driver */
```

Comment:
Do all the architectures support `atomic_int`? Should we really use it in the common code?

Reply:
I found atomic used in BT driver like https://github.com/apache/nuttx/blob/9a28cd2000af50ef15b8ecd08d09a1df34db71da/drivers/wireless/bluetooth/bt_bridge.c#L72

On linux, kernel has `atomic_t` and is widely used, but nuttx doesn't have it. So atomic should be supported by all the arch, but I'm not sure if it's the same on NuttX with atomic_int. Then if we get into trouble later, another atomic type should be present in NuttX and we can just switch a typename.

Reply:
@pkarashchenko here has a default implementation if the toolchain doesn't support atomic operation natively:
https://github.com/apache/nuttx/blob/master/libs/libc/machine/arch_atomic.c

Reply:
@xiaoxiang781216 but this brings a dependency (at least `CONFIG_LIBC_ARCH_ATOMIC` in case if toolchain dose not support for atomic types) into a generic code fir NuttX net implementation

Reply:
@xiaoxiang781216 @pkarashchenko It seems that we need to check both `CONFIG_LIBC_ARCH_ATOMIC` and the toolchain, and cannot have 100% confident that atomic_int is present. So I'd like to use `net_lock` instead (half of the accesses are already inside `net_lock`)

Reply:
In general, I'm not against using atomic variables/operations. Using it may bring certain performance improvements. My concern is that TCP/IP (and netdev interface) stack is shared across all the supported architectures, so maybe it is better to go with implement -> optimize steps here.